trojsten/rules/susi.py

```diff
@@ -0,0 +1,188 @@
+from trojsten.results.generator import CategoryTagKeyGeneratorMixin, ResultsGenerator
+from trojsten.results.representation import ResultsCell, ResultsCol, ResultsTag
+from trojsten.submit.models import Submit
+
+from .default import CompetitionRules
+from .default import FinishedRoundsResultsRulesMixin as FinishedRounds
+
+SUSI_AGAT = "agat"
+SUSI_BLYSKAVICA = "blyskavica"
+SUSI_CIFERSKY_CECH = "cifersky cech"
+
+
+SUSI_AGAT_MAX_COEFFICIENT = 8
+SUSI_ELIGIBLE_FOR_TASK_BOUND = [8, 8, 1000, 1000, 1000, 1000, 1000]
+
+SUSI_CAMP_TYPE = "SuŠi sústredenie"
+
+SUSI_YEARS_OF_CAMPS_HISTORY = 10
+
+
+class SUSIResultsGenerator(CategoryTagKeyGeneratorMixin, ResultsGenerator):
+    def __init__(self, tag):
+        super(SUSIResultsGenerator, self).__init__(tag)
+        self.susi_camps = None
+        self.trojsten_camps = None
+        self.puzzlehunt_participations = None
+        self.has_graduated = None
+        self.coefficients = {}
+
+    def get_user_coefficient(self, user, round):
+        if user not in self.coefficients:
+            if not self.susi_camps:
+                self.prepare_coefficients(round)
+
+            successful_semesters = self.susi_camps.get(user.pk, 0)
+            puzzlehunt_participations = self.puzzlehunt_participations.get(user.pk, 0)
+            trojsten_camps = self.trojsten_camps.get(user.pk, 0)
+            self.coefficients[user] = (
+                5 * successful_semesters + 3 * puzzlehunt_participations + trojsten_camps
+            )
+
+        return self.coefficients[user]
+
+    def get_graduation_status(self, user, round):
+        if user not in self.has_graduated:
+            if not self.has_graduated:
+                self.prepare_coefficients(round)
+
+        return self.has_graduated[user]
+
+    def prepare_coefficients(self, round):
+        """
+        Fetch from the db number of successful semester and number of participation
+        in other Trojsten camps of each user and store them in dictionaries. The prepared
+        data in dictionaries are used to compute the coefficient of a given user.
+        We consider only events happened before given round, so the coefficients are computed
+        correct in older results.
+        """
+
+        self.trojsten_camps = dict(
+            EventParticipant.objects.filter(
+                Q(
+                    # TODO I think commenting next line will retrieve participations in
+                    # all trojsten camps
+                    # event__type__name=SUSI_CAMP_TYPE,
+                    event__end_time__lt=round.end_time,
+                    event__end_time__year__gte=round.end_time.year - SUSI_YEARS_OF_CAMPS_HISTORY,
+                ),
+                Q(going=True) | Q(type=EventParticipant.PARTICIPANT),
+            )
+            .values("user")
+            .annotate(camps=Count("event__semester", distinct=True))
+            .values_list("user", "camps")
+        )
```

Comment:
Myslím si, že toto zaráta aj samotné Suši sústredenia, takže ak niekto bol na Suši sústredku, tak bol úspešný (5 bodov) + bol na sústredku (1 bod) = 6 bodov, čo nechceme.
Nie som si tým úplne istý, neviem do detailov, ako presne to funguje.

Reply:
Asi máš pravdu, najjednoduchšie by sa to asi fixlo pripisovanim 4 bodov za suši sústredko, potom by to bolo 4+1= 5 :D. Alebo teda to nejak odfiltrovať, ale neviem ako.

Reply:
To so 4 bodmi neviem, či bude fungovať, pretože máš ľudí, čo boli úspešní a na sústredku neboli.
Ak boli úspešní v Suši a na sústredko nešli, tak za Suši im chceš dať body skúseností. Ak boli pozvaní na KMS a nešli, tak im nechceš dať body.
Takže si myslím, že pri trojsten_camps chceme dať iba Q(going=True) namiesto celého s "|". Za filter vieš dať exclude, ktorý funguje podobne ako filter, len robí opak, teda budeš mať objects.filter(...).exclude(event__type__name=SUSI_CAMP_TYPE).values... Tým vylúčiš Suši sústredká. Možno to ide aj krajšie.

Reply:
Ešte mi napadlo, že event__end_time__lt=round.end_time môže robiť problém, ak trebárs bude jesenné KSP. Teda riešiteľovi sa medzi sériami zdvihne koeficient, lebo sa mu zaráta to sústredenie (v druhom kole).

Reply:
pravda, tam chceme dat nejak rozumne zaciatok semestra

Reply:
neviem uplne, ako fixnut toto. skusis sa na to pozriet?